src/coreclr/src/gc/gc.cpp

```diff
@@ -9841,44 +10341,406 @@ int gc_heap::object_gennum_plan (uint8_t* o)
+{
+    return heap_segment_demoted_p (region_of (obj));
+}
+
+inline
+void gc_heap::set_region_gen_num (heap_segment* region, int gen_num)
+{
+    heap_segment_gen_num (region) = gen_num;
+}
+
+inline
+void gc_heap::set_region_plan_gen_num (heap_segment* region, int plan_gen_num)
+{
+    int gen_num = heap_segment_gen_num (region);
+    int supposed_plan_gen_num = get_plan_gen_num (gen_num);
+    dprintf (REGIONS_LOG, ("h%d setting plan gen on %Ix->%Ix(was gen%d) to %d(should be: %d) %s",
+        heap_number, (size_t)region, 
+        heap_segment_mem (region),
+        gen_num, plan_gen_num, 
+        supposed_plan_gen_num, 
+        ((plan_gen_num < supposed_plan_gen_num) ? "DEMOTED" : "ND")));
+    if (plan_gen_num < supposed_plan_gen_num)
+    {
+        if (!settings.demotion)
+        {
+            settings.demotion = TRUE;
+        }
+        get_gc_data_per_heap()->set_mechanism_bit (gc_demotion_bit);
+        heap_segment_demoted_p (region) = true;
+    }
+    else
+    {
+        heap_segment_demoted_p (region) = false;
+    }
+
+    heap_segment_plan_gen_num (region) = plan_gen_num;
+}
+#endif //USE_REGIONS
+
+uint8_t* gc_heap::get_uoh_start_object (heap_segment* region, generation* gen)
+{
+#ifdef USE_REGIONS
+    uint8_t* o = heap_segment_mem (region);
+#else
+    uint8_t* o = generation_allocation_start (gen);
+    assert(((CObjectHeader*)o)->IsFree());
+    size_t s = Align (size (o), get_alignment_constant (FALSE));
+    assert (s == AlignQword (min_obj_size));
+    //Skip the generation gap object
+    o += s;
+#endif //USE_REGIONS
+    return o;
+}
+
+uint8_t* gc_heap::get_soh_start_object (heap_segment* region, generation* gen)
+{
+#ifdef USE_REGIONS
+    uint8_t* o             = heap_segment_mem (region);
+#else
+    uint8_t* o             = generation_allocation_start (gen);
+#endif //USE_REGIONS
+    return o;
+}
+
+size_t gc_heap::get_soh_start_obj_len (uint8_t* start_obj)
+{
+#ifdef USE_REGIONS
+    return 0;
+#else
+    return Align (size (start_obj));
+#endif //USE_REGIONS
+}
+
+void gc_heap::clear_gen1_cards()
+{
+#if defined(_DEBUG) && !defined(USE_REGIONS)
+    for (int x = 0; x <= max_generation; x++)
+    {
+        assert (generation_allocation_start (generation_of (x)));
+    }
+#endif //_DEBUG && !USE_REGIONS
+
+    if (!settings.demotion && settings.promotion)
+    {
+        //clear card for generation 1. generation 0 is empty
+#ifdef USE_REGIONS
+        heap_segment* region = generation_start_segment (generation_of (1));
+        while (region)
+        {
+            clear_card_for_addresses (heap_segment_mem (region), heap_segment_allocated (region));
+            region = heap_segment_next (region);
+        }
+#else //USE_REGIONS
+        clear_card_for_addresses (
+            generation_allocation_start (generation_of (1)),
+            generation_allocation_start (generation_of (0)));
+#endif //USE_REGIONS
+    }
+    if (settings.promotion && !settings.demotion)
```

Comment:
Nit: why have a separate if with a swapped but I think identical condition?

Reply:
no reason other than this is a refactoring change and I simply moved the code to be inside a method instead of duplicated in 2 places :) I can definitely make it nicer (including your comment below). I have a cleanup PR coming up so I could do it then unless there's something else I need to change in this PR.

Reply:
Ok.